Bug 1212865 - Add and improve ordering of related instances

jwhitlock opened the pull request:
For API-backed tables with multiple elements, such as [Web/CSS/background](https://developer.mozilla.org/en-US/docs/Web/CSS/background#Browser_compatibility), the order of the table rows is usually different than the source MDN table. This PR implements the features and fixes the bugs so the [importer](https://browsercompat.herokuapp.com/importer/651) will save the order of feature and specification rows ([bug 1212865](https://bugzilla.mozilla.org/show_bug.cgi?id=1212865))

The changes are:
* The importer sets the row orders by specifying ``feature.children`` (order for browser compatibility table) and ``feature.sections`` (order for specification table)
* The API will set the order of related elements on PUT/POST/PATCH, including the other two ordered relations, ``browser.versions`` and ``specification.sections``
* The external tools will set the order of related elements on bulk import
* The API will remember the order in the historical instance
* The instance cache will serialize / deserialize the IDs of related elements in order
* The API will return related items in the order
* The sample JS display code will display rows in that order

Getting all this working required fixing some additional bugs:
* [bug 1159337](https://bugzilla.mozilla.org/show_bug.cgi?id=1159336) - Use the order-aware ``TreeManager`` for features, so that ``feature.children`` are returned in [MPTT](https://github.com/django-mptt/django-mptt) order rather than random MySQL order.
* [bug 1159317](https://bugzilla.mozilla.org/show_bug.cgi?id=1159317) - Fix a bug when doing a PUT to update a view_feature with incomplete data, used by the importer when re-importing a page.
* [bug 1159292](https://bugzilla.mozilla.org/show_bug.cgi?id=1159292) - Attempt to restore related item order when reverting to a past version of an instance.
* A newly discovered bug about instance cache invalidation for common cases such as creating a new instance or closing a changeset.

These changes had a large impact on the documentation / integration tests (specified in docs/doc_cases.json, with results in docs/raw/), which can mostly be ignored outside of running ``make test-integration`` to validate them.

jwhitlock commented:
Rebased on master

groovecoder commented on webplatformcompat/models.py:
```diff
@@ -349,7 +416,8 @@ def post_save_update_cache(sender, instance, created, raw, **kwargs):
             from .tasks import update_cache_for_instance
-            update_cache_for_instance(name, instance.pk, instance, False)
+            update_cache_for_instance(
+                name, instance.pk, instance, update_only=False)
```
Seems to call `update_cache_for_instance` with `update_only=False` a lot ... should we change the default parameter value to `False`?

groovecoder commented on webplatformcompat/models.py:
```diff
@@ -85,12 +96,39 @@ class Feature(MPTTModel):
+                if moved:
+                    next_child.refresh_from_db()
+                if prev_child is None:
+                    next_child.move_to(self)
```
Maybe include the explicit argument value for `move_to`

jwhitlock replied:
Updated to ``first-child``, which is good, because the default was ``last-child``.

groovecoder commented:
Updated, but I don't see a new commit?

jwhitlock commented:
Rebased, and made some changes:
* Used ``next_child.move_to(self, "first-child")`` when placing the first child feature in order
* Changed default to ``update_cache_for_instance(..., update_only=False)``

groovecoder commented on webplatformcompat/tests/test_tasks.py:
```diff
@@ -26,7 +26,7 @@ def test_update_cache_no_invalidation(self):
         update_cache_for_instance('Maturity', self.mat.id)
         self.mock_cache.update_instance.assert_called_once_with(
-            'Maturity', self.mat.id, None, None, update_only=True)
+            'Maturity', self.mat.id, None, None, update_only=False)
```
Still being explicit with these `update_only` argument values even though the parameter default is `False`?

jwhitlock replied:
The mock is for ``cache.update_instance``, which is being called with explicit parameters - see webplatformcompat/tasks.py

groovecoder replied:
Duh. Sorry I didn't notice this was a mock.

jwhitlock commented on webplatformcompat/tasks.py:
```diff
@@ -7,7 +7,7 @@
+        update_only=False):
     cache = Cache()
     invalid = cache.update_instance(
         model_name, instance_pk, instance, version, update_only=update_only)
```
This is what is being mocked in test_tasks.py - explicit arguments and all

groovecoder commented:
Cool, looks good. Good update.
